crates/accelerate/src/circuit_library/iqp.rs

```diff
@@ -0,0 +1,159 @@
+) -> impl Iterator<Item = (StandardGate, SmallVec<[Param; 3]>, SmallVec<[Qubit; 2]>)> + '_ {
+    let num_qubits = interactions.ncols();
+
+    // The initial and final Hadamard layer.
+    let h_layer =
+        (0..num_qubits).map(|i| (StandardGate::HGate, smallvec![], smallvec![Qubit(i as u32)]));
+
+    // The circuit interactions are powers of the CSGate, which is implemented by calling
+    // the CPhaseGate with angles of Pi/2 times the power. The gate powers are given by the
+    // upper triangular part of the symmetric ``interactions`` matrix.
+    let connections = (0..num_qubits).flat_map(move |i| {
+        (i + 1..num_qubits)
+            .map(move |j| (j, interactions[(i, j)]))
+            .filter(move |(_, value)| value % 4 != 0)
+            .map(move |(j, value)| {
+                (
+                    StandardGate::CPhaseGate,
+                    smallvec![Param::Float(PI2 * value as f64)],
+                    smallvec![Qubit(i as u32), Qubit(j as u32)],
+                )
+            })
+    });
+
+    // The layer of T gates. Again we use the PhaseGate, now with powers of Pi/8. The powers
+    // are given by the diagonal of the ``interactions`` matrix.
+    let shifts = (0..num_qubits)
+        .map(move |i| interactions[(i, i)])
+        .enumerate()
+        .filter(|(_, value)| value % 8 != 0)
+        .map(|(i, value)| {
+            (
+                StandardGate::PhaseGate,
+                smallvec![Param::Float(PI8 * value as f64)],
+                smallvec![Qubit(i as u32)],
+            )
+        });
+
+    h_layer
+        .clone()
+        .chain(connections)
+        .chain(shifts)
+        .chain(h_layer)
+}
+
+/// This generates a random symmetric integer matrix with values in [0,7].
+fn generate_random_interactions(num_qubits: u32) -> Array2<i64> {
+    // We first generate all unique values, which can be thought of generating the upper
+    // triangular matrix. This contains N(N+1)/2 unique random values, stored as
+    //    [v_0 v_1 ... v_{N-1}]
+    //    [    v_N ... v_{2N-2}]
+    //             ...
+    //    [            v_{-1}]
+    let num_qubits = num_qubits as usize;
+    let num_values = num_qubits * (num_qubits + 1) / 2;
+    let values: Vec<i64> = (0..num_values)
+        .map(|_| rand::thread_rng().gen_range(0..8) as i64)
+        .collect();
+
+    // We then build the matrix of dimension NxN by reading from the ``values`` vector.
+    // * since the matrix is symmetric, we treat (i, j) and (j, i) equally, which we implement
+    //   here by sorting the indices before accessing the value vector
+    // * in each row, the offset with which we access ``values`` changes: in row K, we need to
+    //   start reading from index: \sum_{k=0}^K N-k
+    Array2::from_shape_fn((num_qubits, num_qubits), |(i, j)| {
+        let (low, high) = if i < j { (i, j) } else { (j, i) };
+        let offset = low * (2 * num_qubits - low + 1) / 2;
+        values[offset + high - low]
+    })
```

Comment:
Out of curiosity, is this faster than a more naive approach in https://github.com/Qiskit/qiskit/blob/fbfe7388f3efdaaeb70dc96aa35f649c1897c5d4/crates/accelerate/src/synthesis/clifford/random_clifford.rs#L43-L53.

Reply:
and a follow-up question: where are the values below the diagonal even used?

Reply:
* I used the `from_shape_fn` to avoid the matrix initialization with zeros, before we overwrite the values later 🙂 but I can benchmark the difference (I assume it is minimal).
* They are not, see also this comment: https://github.com/Qiskit/qiskit/pull/13241/files/84d92bc2e9ba2a3089ed8126481744d803197429#r1790125684 But since we have to fill the matrix with some values, I'm generating the full, valid IQP matrix here

Reply:
In fact, the "naive" approach is even faster, I guess because it can omit the check "i < j"? 🤔  Here's the times (in `s`) to create random IQP circuits with `size` qubits
```
size   from_shape_fn   for-loops
----------------------------------
100    0.001           0.001
500    0.012           0.011
1000   0.057           0.048
5000   2.613           2.405
```
I'll change to the for-loops approach 🙂

Reply:
Added in 1e103807e